### DISCUSSION TITLE
How to serve ML binary efficiently

### DISCUSSION THREAD

**Discussion by gamamoe** [fastapi/fastapi#2695] [Questions] [answered]
Hello guys, I'm new to use FastAPI to build API server predicting the result from ML model.
From the useful and kind tutorial, I just learned about basic usage of FastAPI.
But, I'm entry-level engineer to build API server and want to clear about something.

In my case, there is ML train app which is independent API server.
It saves trained binary in specific path every day, such as 
```
/home/foo/trained_model/20210120.bin
/home/foo/trained_model/20210121.bin
/home/foo/trained_model/20210122.bin
...
```

What I want to do is 

1) When I execute API server, it loads latest trained binary and reuses it before achieving new model binary
2) While running API server, it detects newly trained binary and switch it to serve

When I searched above context in Google, reusing object could be achieved by global object (maybe?)
But, I'm wondering how to reload model automatically and efficiently.

Thanks in advance about your opinions!

**Comment by sm-Fifteen** [comment, marked as answer]
You basically want to always serve the most recent file in that directory? Like having a `myapp:80/trained_model/latest` route that applications can try to access knowing it's always up to date?

**Reply by gamamoe** [reply]
Hi @sm-Fifteen Thanks for comment!

> You basically want to always serve the most recent file in that directory?

Yes exactly, another batch app makes and stores model binary to specific path every day.

> Like having a myapp:80/trained_model/latest route that applications can try to access knowing it's always up to date?

I'm wondering the meaning of this sentence, could you explain more detail?

**Reply by sm-Fifteen** [reply]
> I'm wondering the meaning of this sentence, could you explain more detail?

You often see file shares that have multiple versions of the same thing (let's say python versions) have a URL structure like this:

```
/programs
  /python
    /python-3.6.zip
    /python-3.7.zip
    /python-3.8.zip
    /python-3.9.zip
    /latest
```

With `/latest` automatically pointing towards the most recent version, so you don't have to keep track of new releases all the time.

Your file names conveniently happen to have the date in YYYYMMDD format at the start of their name, so sorting them alphabetically is equivalent to sorting them chronologically. That means you could probably just do something like this.

```py
import os

MODEL_DIR = '/home/foo/trained_model/'

@app.get('/dolpang2/ml_model/latest')
def latest_model():
    
    with os.scandir(MODEL_DIR) as scanner:
        latest_path = None
        latest_name = ""
    
        for entry in scanner:
            if not entry.name.startswith('.') and entry.is_file() and entry.name > latest_name:
                latest_name = entry.name
                latest_path = entry.path

    if latest_path is None:
        ... # No suitable file found, error 404 or 406

    # open latest_path and serve its content
```

**Reply by gamamoe** [reply]
@sm-Fifteen Sorry for late reponse : ( Thanks for kind explanation! I understand now with your example. But I have one more question, it seems reading latest model binary in get method (/dolpang2/ml_model/latest) is required request-level file-io (each request, worker need to file-io). Is it ok for serving large traffic? Or is there other approach like worker-level file-io? Thanks in advance!

**Reply by sm-Fifteen** [reply]
Well, that's the obvious solution for doing this on the FastAPI side of things. Supposedly `os.scandir` is extremely fast, since it doesn't actually open the files, only the filesystem metadata tables, so while performance *might* be suboptimal, it probably wouldn't be too bad, although you might want to test this yourself rather than taking my word for it. While you could have a filesystem watcher like python-watchdog be initialized alongside your FastAPI app so it runs in the same thread and updates some global variable that the worker can then check for, that's not something I have experience with and I wouldn't feel comfortable recommending it.

That's if you really want to have FastAPI handle the decision of what file to serve, anyway. What I would probably do in a similar situation is have `latest` be a symbolic link pointing to the most recent file, and have FastAPI (or a dedicated HTTP static file server like Apache or Nginx) serve it as if it were a normal file. You could then just have your filesystem watcher run as a separate service and make it update the symbolic link whenever it detects a new file has been added to the `ml_model` directory.

**Reply by gamamoe** [reply]
@sm-Fifteen omg, Really appreciated it! I may need to study more about back-end development and your comment would be really helpful what should I have to learn more about. I will follow up based on your comment. Really thx!
